modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/ComposerServiceImpl.java

```diff
@@ -1091,8 +1091,18 @@ public Job image(Track sourceTrack, String profileId, Map<String, String> proper
         params.put("track-duration", sourceTrack.getDuration().toString());
         params.put("time", Double.toString(time));
         incident().recordFailure(job, IMAGE_EXTRACTION_TIME_OUTSIDE_DURATION, params);
-        throw new EncoderException("Can not extract an image at time " + time + " from a track with duration "
-                + sourceTrack.getDuration());
+
+        // Format duration (a `long`, in milliseconds) as seconds, with a fractional part.
+        // Why not just cast to `double` and divide by 1000? Because of precision loss,
+        // both during the cast and the division.
+        String duration = Long.toString(sourceTrack.getDuration());
```

Comment:
Please remove the comment and replace the code by something like:

```
Duration duration = Duration.ofMillis(sourceTrack.getDuration());
throw new EncoderException("Cannot extract image at position " + time + " second(s) from a track with the duration "
                + duration.getSeconds() + "." + duration.minusSeconds(duration.getSeconds()).toMillis() + "second(s)");
```

Admittedly still quite ugly as Java 8 does not yet support the toMillisPart() method introduced later in Java 9.

Also: This is just the exception. It might be preferable to (also) adjust the incident IMAGE_EXTRACTION_UNKNOWN_DURATION

Reply:
@staubesv thanks for the chance to finally take a look at the "new" time API :smile: I fixed that part and added 0-padding, which I also forgot. :wink:

As for the incident message: I'm not quite sure what you mean. First, you probably mean `IMAGE_EXTRACTION_TIME_OUTSIDE_DURATION`, not `IMAGE_EXTRACTION_UNKNOWN_DURATION`, as the latter is for a different error, right? But even then: `IMAGE_EXTRACTION_TIME_OUTSIDE_DURATION` ist just a number (`17` to be precise). Where can I find the message you are talking about?

Reply:
@JulianKniephoff I still don't get what it means to  "Extract an image [...] into a second track":

> Can not extract an image from " + time + " second(s) into a " + formattedDuration + "second track

I think that should be something like: "Cannot extract image at position " + time + " second(s) " from a track with duration " + formattedDuration

The incident is supposed to be the user-readable error message an will be displayed in the Admin UI. You can find the incident messages for the composer service in https://github.com/opencast/opencast/tree/develop/modules/composer-ffmpeg/src/main/resources/incident-l10n

Reply:
The message is not "extract an image [...] into a second track", but something like "extract an image from 6 seconds into a 5 second track."

I'm no native speaker, of course, but as far as I know "n <units> into an m <unit> thing" is a perfectly valid phrasing of the two facts that the thing is m units long and that we are talking about the point in time n units from the start of the thing.

If you insist, though, I will change it of course. :wink:

The incident message should now get the formatted duration as well.

Reply:
@JulianKniephoff What confuses me about the message is how can you extract an image into a track? It is extracted from a track, isn't it?

Reply:
It is extracted **from** some time into the track. The "into" belongs to "some time" into this case, not to the verb. "n seconds into the track" specifies a point in time, relative to the start of the track.

Reply:
@JulianKniephoff I'm also not a native speaker, but I don't think the current wording makes too much sense.

@karendolan As you are a native speaker, could you please have a quick look at the error message under discussion and suggest an understandable and correct formulation?

Reply:
Okay so apparently Google Groups thinks I'm a spam bot, so I can't answer on list right now, unfortunately.

Here is what I drafted (slightly adapted to GitHub)

> Oh no, this got out of control, I'm sorry! ^.^"
>
> @smarquard is right, an error message should be clear to the users who have to read and act on it in the end, regardless of whether it is correct and/or natural. Seeing as this made it to list this is not the case here, so it was a bad choice on my part. I will change it.
>
> But I have to say, @staubesv, it feels good to be backed up by native speakers on this (thanks, @JamesUoM!), so I guess in the end everyone is happy. ;D

I took a hint from the incident message now, which is very clear in my opinion, especially after adding units to it as well.

Can we merge this now? :see_no_evil: :hear_no_evil: :speak_no_evil:

Reply:
@JulianKniephoff Well... seems I need some English lessons ;-)

I'm on it.